### PR TITLE
add a Tower subgraph service

### PR DESCRIPTION
**Pull request by Geal**
we already use Tower everywhere and a hyper client can be used as a
service. This can be more flexible than using reqwest because we will
not need the multiple steps of URL manipulation, and opentelemetry
header propagation will integrate better because it would fall under the
apollo-router-core log filter

### CODE REVIEW

**Comment by Geal** [comment]
I just verified that header propagation works properly

**Comment by Geal** [comment]
perf test please

**Comment by Geal** [comment]
since CI benchmarks are a bit unreliable, I did a small benchmark locally.

2 cores, 1000 clients, 100000 reqs

# main bfcd8071

Summary:
  Total:        34.3817 secs
  Slowest:      3.4497 secs
  Fastest:      0.0059 secs
  Average:      0.3209 secs
  Requests/sec: 2908.5247
  
  Total data:   1192300000 bytes
  Size/request: 11923 bytes

Response time histogram:
  0.006 [1]     |
  0.350 [67561] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.695 [14191] |■■■■■■■■
  1.039 [13020] |■■■■■■■■
  1.383 [3123]  |■■
  1.728 [1484]  |■
  2.072 [443]   |
  2.417 [135]   |
  2.761 [30]    |
  3.105 [9]     |
  3.450 [3]     |


Latency distribution:
  10% in 0.0660 secs
  25% in 0.0786 secs
  50% in 0.0981 secs
  75% in 0.5999 secs
  90% in 0.8352 secs
  95% in 1.0627 secs
  99% in 1.5827 secs
  
  # geal/tower-client branch

Summary:
  Total:        28.6684 secs
  Slowest:      2.9447 secs
  Fastest:      0.0052 secs
  Average:      0.2678 secs
  Requests/sec: 3488.1663
  
  Total data:   1192300000 bytes
  Size/request: 11923 bytes

Response time histogram:
  0.005 [1]     |
  0.299 [67326] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.593 [15953] |■■■■■■■■■
  0.887 [11601] |■■■■■■■
  1.181 [3147]  |■■
  1.475 [1504]  |■
  1.769 [266]   |
  2.063 [174]   |
  2.357 [20]    |
  2.651 [7]     |
  2.945 [1]     |


Latency distribution:
  10% in 0.0559 secs
  25% in 0.0664 secs
  50% in 0.0820 secs
  75% in 0.4943 secs
  90% in 0.6779 secs
  95% in 0.9016 secs
  99% in 1.3104 secs

File: apollo-router/src/executable.rs [hunk 1]
```diff
@@ -26,7 +26,7 @@ pub struct Opt {
     /// Log level (off|error|warn|info|debug|trace).
     #[clap(
         long = "log",
-        default_value = "apollo_router=info,router=info,apollo_router_core=info,apollo_spaceport=info,tower_http=info,reqwest_tracing=info",
+        default_value = "apollo_router=info,router=info,apollo_router_core=info,apollo_spaceport=info,tower_http=info",
```
Comment (garypen) [inline apollo-router/src/executable.rs:29]: Do we need to add `hyper=info` ?

**Review by garypen — APPROVED** [review]

File: apollo-router-core/src/services/tower_subgraph_service.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            request.headers_mut().insert(ACCEPT, app_json);
+
+            global::get_text_map_propagator(|injector| {
+                injector.inject_context(
+                    &Span::current().context(),
+                    &mut RequestInjector {
+                        request: &mut request,
+                    },
+                )
+            });
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router-core/src/services/tower_subgraph_service.rs:68]: yes that's why. But there's a nice side effect here: now this is part of `apollo-router-core`, so if the log filters only include `apollo_router` and `apollo_router_core`, header propagation will still work

File: apollo-router/src/executable.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/executable.rs]: done in 2d94917a. I see the same spans generated

**Review by o0Ignition0o — APPROVED** [review]

**Pull request merged**